url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/SpecialOperations.java

```diff
@@ -0,0 +1,607 @@
+					s.transitions.add(new Transition(t.min, t.max, q));
+		b.deterministic = true;
+		b.removeDeadTransitions();
+		return b;
+	}
+
+	/**
+	 * Returns an automaton that accepts the trimmed language of the given
+	 * automaton. The resulting automaton is constructed as follows: 1) Whenever
+	 * a <code>c</code> character is allowed in the original automaton, one or
+	 * more <code>set</code> characters are allowed in the new automaton. 2)
+	 * The automaton is prefixed and postfixed with any number of
+	 * <code>set</code> characters.
+	 *
+	 * @param set set of characters to be trimmed
+	 * @param c   canonical trim character (assumed to be in <code>set</code>)
+	 */
+	public static DefaultAutomaton trim(DefaultAutomaton a, String set, char c) {
+		a = a.cloneExpandedIfRequired();
+		State f = new State();
+		addSetTransitions(f, set, f);
+		f.accept = true;
+		for (State s : a.getStates()) {
+			State r = s.step(c);
+			if (r != null) {
+				// add inner
+				State q = new State();
+				addSetTransitions(q, set, q);
+				addSetTransitions(s, set, q);
+				q.addEpsilon(r);
+			}
+			// add postfix
+			if (s.accept != null)
+				s.addEpsilon(f);
+		}
+		// add prefix
+		State p = new State();
+		addSetTransitions(p, set, p);
+		p.addEpsilon(a.initial);
+		a.initial = p;
+		a.deterministic = false;
+		a.removeDeadTransitions();
+		a.checkMinimizeAlways();
+		return a;
+	}
+
+	private static void addSetTransitions(State s, String set, State p) {
+		for (int n = 0; n < set.length(); n++)
+			s.transitions.add(new Transition(set.charAt(n), p));
+	}
+
+	/**
+	 * Returns an automaton that accepts the compressed language of the given
+	 * automaton. Whenever a <code>c</code> character is allowed in the
+	 * original automaton, one or more <code>set</code> characters are allowed
+	 * in the new automaton.
+	 *
+	 * @param set set of characters to be compressed
+	 * @param c   canonical compress character (assumed to be in <code>set</code>)
+	 */
+	public static DefaultAutomaton compress(DefaultAutomaton a, String set, char c) {
+		a = a.cloneExpandedIfRequired();
+		for (State s : a.getStates()) {
+			State r = s.step(c);
+			if (r != null) {
+				// add inner
+				State q = new State();
+				addSetTransitions(q, set, q);
+				addSetTransitions(s, set, q);
+				q.addEpsilon(r);
+			}
+		}
+		// add prefix
+		a.deterministic = false;
+		a.removeDeadTransitions();
+		a.checkMinimizeAlways();
+		return a;
+	}
+
+	/**
+	 * Returns an automaton where all transition labels have been substituted.
+	 * <p/>
+	 * Each transition labeled <code>c</code> is changed to a set of
+	 * transitions, one for each character in <code>map(c)</code>. If
+	 * <code>map(c)</code> is null, then the transition is unchanged.
+	 *
+	 * @param map map from characters to sets of characters (where characters
+	 *            are <code>Character</code> objects)
+	 */
+	public static DefaultAutomaton subst(DefaultAutomaton a, Map<Character, Set<Character>> map) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.